Add https support

turingmachinerepair opened the pull request:
This PR introduces:
1. Support for https scheme for uri_map requests.
2. Gun version bump from 1.3.2 to 1.3.3 to fix cowlib version mismatch. The mismatch caused https test to fail.

arentrue commented on src/hapi.erl:
```diff
@@ -37,7 +37,9 @@
--type uri() :: {http, http_uri:user_info(),
+-type scheme() :: http | https.
+-type transport() :: tcp | tls.
+-type uri() :: {http | https, http_uri:user_info(),
```
```suggestion
-type uri() :: {scheme(), http_uri:user_info(),
```

turingmachinerepair replied:
Done.

arentrue commented on src/hapi.erl:
```diff
@@ -144,11 +146,12 @@ req(Method, URI0, Opts) when is_map(URI0), map_size(URI0) > 0 ->
     Host = maps:get(host, URI),
     Port = maps:get(port, URI, 80),
     Path = maps:get(path, URI, ""),
+    Scheme = list_to_existing_atom(maps:get(scheme, URI)),
```
According to https://www.erlang.org/doc/man/uri_string.html#type-uri_map spec the `scheme` is `unicode:chardata()` meaning it can be a _binary_ as well. Since we only support _http_ and _https_ here, I would map the scheme explicitly:
```erlang
format_scheme("http") -> http;
format_scheme(<<"http">>) -> http;
format_scheme("https") -> https;
format_scheme(<<"https">>) -> https.
```

turingmachinerepair replied:
Done

arentrue commented on src/hapi.erl:
```diff
@@ -307,6 +315,7 @@ close(AddrPort, ConnPid, StreamRef, Req) ->
         false -> gun:close(ConnPid)
     end.
 
+
```
Irrelevant diff

turingmachinerepair replied:
Done

arentrue replied:
mm, it seems additional line is still here.

turingmachinerepair replied:
I looked at a wrong line. Now fixed.

arentrue commented on test/hapi_eunit.erl:
```diff
@@ -28,6 +28,11 @@ get_test() ->
+
 get_uri_string_test() ->
     Host = "127.0.0.1", Port = get_port(), Path = "/empty",
     URL = "http://" ++ Host ++ ":" ++ integer_to_list(Port) ++ Path,
```
Should we test https and a binary scheme here too?

turingmachinerepair replied:
Added test for binary URL and `uri_map` with binary fields. The `uri_map` test also uses https.

Commit:
BP-112: Add https support

Also:
1. Bump gun version to 1.3.3 to fix cowlib version mismatch with cowboy. The mismatch caused https tests to fail instantly.
2. Bump cowboy version to 2.8.0 and fix hapi_httpd_test reponse with code 204
3. Add extra tests to check binary URLs and uri_maps with binary fields
4. Bump jiffy version to 1.1.1 to fix compatibility issues with OTP 23+
5. Update readme
